src/main/java/com/fittracker/fittracker/service/ProductService.java

```diff
@@ -33,4 +36,30 @@ public ProductResponse findById(UUID id) {
             .map(ProductResponse::fromProduct)
             .orElseThrow(() -> new ProductNotFoundException(id));
     }
+
+    @Transactional
+    public ProductResponse update(UUID id, ProductRequest productRequest) {
+        Product dbProduct = productRepository.findByIdAndUserIdAndActive(
+                id, SecurityHelper.getUserId(), true)
+            .orElseThrow(() -> new ProductNotFoundException(id));
+
+        Product newProduct = productRequest.toProduct();
+
+        if (dbProduct.equals(newProduct)) {
+            throw new ProductAlreadyExistsException();
+        }
+
+        dbProduct.setActive(false);
+        productRepository.save(dbProduct);
+
+        newProduct.setId(id);
+        newProduct.setUserId(SecurityHelper.getUserId());
+        newProduct.setVersion(dbProduct.getVersion() + 1);
+
+        Product product = productRepository.saveNew(newProduct)
+            .orElseThrow(() -> new TransactionSystemException("Cannot find updated product"));
```

Comment:
Let's create our own exception for that. Also, it can occur when insert itself fails, not only when we fail to find the product.